gcc/rust/resolve/rust-name-resolver.h

```diff
@@ -202,6 +205,9 @@ class Resolver
   std::map<NodeId, NodeId> resolved_labels;
   std::map<NodeId, NodeId> resolved_macros;
 
+  // misc
+  std::map<NodeId, NodeId> misc_resolved_items;
```

Comment:
What is considered a misc item?

Reply:
This is a bit of a hack, basically in our name resolver we have the different scopes and the original idea I had back last year was well most of the time we can resolve things directly in the name resolver (which turned out not to be the case for any Path items). So when we do resolve things we do:

```
insert_resolved_{type,name,label,macro} reference_node_id -> definition_node_id
```

We use these mappings during type checking and compilation to figure things out to NodeId's then use mappings to get over to HirId's to then finally query things.

The issue is for this bug in https://github.com/Rust-GCC/gccrs/pull/1545/commits/b17e69852ea999547305b19e9c8da385f6c31e1e the enum is declared in the block scope and these are handled slightly differently during name resolution.

During name resolution we have the top-level scan then the deep scan. The top-level scan puts all top-level items directly into the uppermost scope rib but for function items we don't drill into these until the 2nd pass during name resolution. This means the declared names are actually existing in nested scopes as you would expect.

We guard this behaviour with calls to append_reference_to_decl in the name resolver which means when you do call inser_resolved_name it looks up in reverse order of the ribs to add a mapping of this node id maps back to this declaration defined in this rib. I though this would be more useful but turns out its just a bit of a pain. The only good part of this assertion is it keeps things honest right now.

This misc_resolved_items is to handle the case of items that are defined within nested scopes. Its a catch all so when the analysis passes lookup resolved node_ids they will always be able to lookup in order of:

name->type->label->misc

So long as no errors have occurred. Using the name resolver information for HIR stuff during lint passes is probably not the best solution but its the only interface we have right now.